Implement Kitty Color Protocol (OSC 21)

jcollie opened the pull request:
Kitty 0.36.0 added support for a new OSC escape sequence for quering, setting, and resetting the terminal colors. Details can be found [here](https://sw.kovidgoyal.net/kitty/color-stack/#setting-and-querying-colors).

This fully parses the OSC 21 escape sequences, but only supports actually querying and changing the foreground color, the background color, and the cursor color because that's what Ghostty currently supports. Adding support for the other settings that Kitty supports changing ranges from easy (cursor text) to difficult (visual bell, second transparent background color).

jcollie committed:
Implement Kitty Color Protocol (OSC 21)

Kitty 0.36.0 added support for a new OSC escape sequence for
quering, setting, and resetting the terminal colors. Details
can be found [here](https://sw.kovidgoyal.net/kitty/color-stack/#setting-and-querying-colors).

This fully parses the OSC 21 escape sequences, but only supports
actually querying and changing the foreground color, the background
color, and the cursor color because that's what Ghostty currently
supports. Adding support for the other settings that Kitty supports
changing ranges from easy (cursor text) to difficult (visual bell,
second transparent background color).

jcollie committed:
fix test error and improve error logging

mitchellh reviewed (changes requested):
Very excellent work! Truly! I've noted some changes I'd like to see. Thank you!

jcollie committed:
Address review comments.

- Cap the total number of requests at twice the maximum number of
  keys (currently 263, so 526 requests). Basically you can set and then
  query every key in one message. This is an absurdly high number
  but should prevent serious DOS attacks.
- Clarify meaning of new hex color codes.
- Better handle sending messages to the renderer in a way that should
  prevent deadlocks.
- Handle 0-255 palette color requests by creatively using non-exhautive
  enums.
- Fix an error in the query reply.

mitchellh reviewed (commented):
Thanks! I want to play around with one thing but the rest looks good.

mitchellh committed:
terminal: make kitty color tests a bit more readable

mitchellh committed:
terminal: stylistic tweaks to kitty color protocol parsing

mitchellh committed:
terminal: formatter for kitty color protocol kinds

mitchellh committed:
termio: use new formatter, palette helper for kitty color protocol

mitchellh reviewed (changes requested):
Couple more changes requested. I can get to them soon or you can, either way. They're small.

mitchellh committed:
termio: move renderer message sending out to separate func

mitchellh committed:
terminal: move kitty color structs out to kitty package
